python/tests/core/test_delta_engine.py

```diff
@@ -0,0 +1,296 @@
+        mocker.patch.object(engine, "_write_delta_rs_dataset", return_value=mock.Mock())
+        result = engine.save_delta_fg(dataset=mock.Mock(), write_options=None, validation_id=None)
+        assert mock_commit.called
+        assert result == mock_commit.return_value
+
+    def test_delete_record_importerror_spark_delta_spark_missing(self, mocker):
+        _patch_client(mocker, is_external=False)
+        spark = mock.Mock()
+        fg = _make_fg("hopsfs://nn:8020/p")
+        engine = DeltaEngine(1, "fs", fg, spark, None)
+        with pytest.raises(ImportError) as e:
+            engine.delete_record(delete_df=mock.Mock())
+        assert "delta-spark" in str(e.value)
+
+    def test_delete_record_importerror_rs_deltalake_missing(self, mocker):
+        _patch_client(mocker, is_external=False)
+        fg = _make_fg("hopsfs://nn:8020/p")
+        engine = DeltaEngine(1, "fs", fg, None, None)
+        with pytest.raises(ImportError) as e:
+            engine.delete_record(delete_df=mock.Mock())
+        assert "hops-deltalake" in str(e.value)
+
+    def test_write_delta_dataset_importerror_missing_delta_spark(self, mocker):
+        _patch_client(mocker, is_external=False)
+        spark = mock.Mock()
+        fg = _make_fg("hopsfs://nn:8020/p")
+        engine = DeltaEngine(1, "fs", fg, spark, None)
+        with pytest.raises(ImportError) as e:
+            engine._write_delta_dataset(dataset=mock.Mock(), write_options=None)
+        assert "delta-spark" in str(e.value)
+
+    def test_write_delta_rs_dataset_importerror_missing_deltalake(self, mocker):
+        _patch_client(mocker, is_external=False)
+        fg = _make_fg("hopsfs://nn:8020/p")
+        engine = DeltaEngine(1, "fs", fg, None, None)
+        with pytest.raises(ImportError) as e:
+            engine._write_delta_rs_dataset(dataset=mock.Mock())
+        assert "hops-deltalake" in str(e.value)
+
+    def test_prepare_df_for_delta_importerror(self, monkeypatch):
+        # Force ImportError by ensuring pandas/pyarrow imports fail
+        import builtins
+        original_import = builtins.__import__
+
+        def fake_import(name, *args, **kwargs):
+            if name in ("pandas", "pyarrow"):
+                raise ImportError("missing")
+            return original_import(name, *args, **kwargs)
+
+        monkeypatch.setattr(builtins, "__import__", fake_import)
+        with pytest.raises(ImportError):
+            DeltaEngine._prepare_df_for_delta(df=mock.Mock())
+
+    def test_vacuum_executes_sql(self, mocker):
+        _patch_client(mocker, is_external=False)
+        spark = mock.Mock()
+        fg = _make_fg("hopsfs://nn:8020/p")
+        engine = DeltaEngine(1, "fs", fg, spark, None)
+        fg.prepare_spark_location.return_value = "/loc"
+        engine.vacuum(24)
+        spark.sql.assert_called_once()
+        assert "VACUUM '/loc' RETAIN 24 HOURS" in spark.sql.call_args[0][0]
```

Comment:
just a suggestion to improve readability would be to separate the steps in tests, like done in other places. This would allow to know at a glance what is tested (here you have to read the code and extrapolate)

Arrange, Act, Assert